Nixnerd/expand object support

nixnerd2038 opened the pull request:
Hey dude, I know the logic's a lil crusty, but it works well, and at least we get more than records out of it for CRUD. Let me know what you think. I ran a test suite for the string interpolation and all of them filter as expected. If you want I can beef em up and add the tests and some tox config. I'll test the interaction of the object's that weren't supported against Infoblox itself at work tomorrow. I did grab the object names from my local WAPI instance from ?_schema, so they're all legit. Did some linting and doc strings while I was in there, too. We use your lib on my team to host a Swagger API for DNS/DDI. I figured two years was too long to go without an update, so gave you some free spit n polish for the tears you've saved me over the years...Cheers.

Commit:
add logic to __getattr__ that supports parent-child objects with colon delimeter

Commit:
add logic to __getattr__ that supports parent-child objects with colon delimeter

Commit:
some linting

Commit:
some comments

nixnerd2038 commented:
I could also make it less deterministic by passing the InfobloxWAPI object to the __getattr__ method, and populating the root_objects with a _schema call.

dnaeon commented on pyinfoblox/__init__.py:
```diff
@@ -68,19 +77,45 @@ def __getattr__(self, attr):
-            attr = attr.replace('_', ':', 1)
+
+        # all root objects that have children, according to WAPI v2.9
+        root_objects = ['certificate','ciscoise','ddns','dhcp','discovery',\
```
No need for backslashes at the end of each line.

nixnerd2038 replied:
Fixed. Bad habit.

dnaeon commented:
Looks great, thanks! If you have the tests, please include them in the PR as well.

Thanks again!

Commit:
remove backslashes

nixnerd2038 commented:
> Looks great, thanks! If you have the tests, please include them in the PR as well.
> 
> Thanks again!

Do you have a preferred framework? I'm used to tox, but I can simplify with an assert script.

dnaeon commented:
> > Looks great, thanks! If you have the tests, please include them in the PR as well.
> > Thanks again!
> 
> Do you have a preferred framework? I'm used to tox, but I can simplify with an assert script.

These days I tend to use `pytest`, but I don't have any strong preference. If tox is what you are used to, feel free to go with it.

nixnerd2038 committed:
add tests, start

dnaeon commented on tests/__init__.py:
```diff
@@ -0,0 +1,20 @@
+import requests
```
Doesn't seem to be used at all, so this could be removed.

dnaeon commented on tests/__init__.py:
```diff
@@ -0,0 +1,20 @@
+                 wapi='https://localhost:8080/wapi/v1.1/',
+                 verify=False)
+        
+        with open('./test_data/host.json') as jsonfile:
```
This won't work, since the file will be closed once done with it. The same issue applies to the other `with open ...` statements.

nixnerd2038 replied:
Reading and assigning it to a class var, that's then inherited preserves the var scope, so this does work. 
https://stackoverflow.com/questions/45100271/scope-of-variable-within-with-statement

dnaeon replied:
I was referring to the fact that you don't read anything from the file within the context manager, so basically the `host_base` property now points to a closed handle.


```python

>>> class Foo:
...     def __init__(self):
...         with open('/etc/passwd') as f:
...             self.some_file = f
...
>>>>>> class Bar(Foo):
...     def __init__(self):
...         super().__init__()
...
>>>
>>> bar = Bar()
>>> bar.some_file
<_io.TextIOWrapper name='/etc/passwd' mode='r' encoding='UTF-8'>
>>>
>>> bar.some_file.readlines()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: I/O operation on closed file.
```

nixnerd2038 replied:
Yeah that was dumb, forgot I hadn't pushed the json.loads(jsonfile)

dnaeon commented on tests/host_tests.py:
```diff
@@ -0,0 +1,86 @@
+        Returns:
+            res (dict): The reponse dict
+        """
+        res = infoblox.record_host.get(name=fqdnn)
```
Typo in `name=fqdnn`, should be `name=fqdn`.

dnaeon commented on tests/host_tests.py:
```diff
@@ -0,0 +1,86 @@
+            res (dict): The reponse dict
+        """
+        res = infoblox.record_host.create(name=fqdn,
+                                          ipv4addrs=[{ip_address}],
```
Does this need to be a `set`? You are already specifying this as a list, so no need to pass a `set` as well.

dnaeon reviewed (changes requested):
I would also recommend checking out [vcrpy](https://vcrpy.readthedocs.io/en/latest/), which along with `pytest` would be better IMO.

Let me know what you think.

Thanks!

nixnerd2038 commented:
Still in flight, want me to close the PR til I'm done?

Commit:
add static host tests

Commit:
add data files: network, network_update, network_container, network_container_update

Commit:
fix data loading: add json loader to file handler context, add class inheritence

Commit:
add tests: crud hosts, get_next_ip

Commit:
add tests: crud hosts, get_next_ip

Commit:
add tests: crud networks

Commit:
add tests: crud network containers

Commit:
abstract root_objects to grid call

Commit:
finish meta test for object name validity

Commit:
finish meta test for object name validity

Commit:
add tests: crud hosts, get_next_ip

Commit:
add teardowns and network init

Commit:
update networks

Commit:
add stup, teardown methods; add test_grid

Commit:
abstract root_objects to wapi call for version independance

Commit:
add requirements and tox for test venv runner

nixnerd2038 committed:
fix data loading: add json loader to file handler context, add class inheritence

nixnerd2038 committed:
add stup, teardown methods; add test_grid

nixnerd2038 committed:
finish meta test for object name validity

nixnerd2038 committed:
some linting

nixnerd2038 committed:
add tests: crud hosts, get_next_ip

nixnerd2038 committed:
add tests: crud hosts, get_next_ip

nixnerd2038 committed:
add teardowns and network init

nixnerd2038 committed:
add logic to __getattr__ that supports parent-child objects with colon delimeter

nixnerd2038 committed:
add requirements and tox for test venv runner

nixnerd2038 committed:
finish meta test for object name validity

nixnerd2038 committed:
some comments

nixnerd2038 committed:
update networks

nixnerd2038 committed:
abstract root_objects to grid call

nixnerd2038 committed:
add tests, start

nixnerd2038 committed:
abstract root_objects to wapi call for version independance

nixnerd2038 committed:
add data files: network, network_update, network_container, network_container_update

nixnerd2038 committed:
add tests: crud networks

nixnerd2038 committed:
remove backslashes

nixnerd2038 committed:
add tests: crud network containers

nixnerd2038 committed:
add tests: crud hosts, get_next_ip

nixnerd2038 committed:
add logic to __getattr__ that supports parent-child objects with colon delimeter

nixnerd2038 committed:
add static host tests

nixnerd2038 commented:
It's almost done, still need to add VCR, some next_available tests for network/networkcontainer, and spruce up some of the cludgy network setup in the host/a/ptr tests. Sorry for the force push, wanted to get my work email out of the git logs. Data's the same, just rebased. Let me know if you like that root_objects call in __getattr__, the "exclusions" follow a similar pattern [x[-1] for x in root_objects if ':' in x and '_' in x.split(':')[-1]]. And sorry it's taking so long, been sick the past week.

dnaeon commented:
> It's almost done, still need to add VCR, some next_available tests for network/networkcontainer, and spruce up some of the cludgy network setup in the host/a/ptr tests. Sorry for the force push, wanted to get my work email out of the git logs. Data's the same, just rebased. Let me know if you like that root_objects call in **getattr**, the "exclusions" follow a similar pattern [x[-1] for x in root_objects if ':' in x and '_' in x.split(':')[-1]]. And sorry it's taking so long, been sick the past week.

No worries, take your time! No need for rushing.

Hope you get better soon! :)

Commit:
update libs

Commit:
cleanup comments

Commit:
fix commenting
